File: ArchiveProcessor/ArchiveCapture/app/src/main/java/com/archiveprocessor/capture/net/DriveRelayTransport.kt
package com.archiveprocessor.capture.net

import org.json.JSONObject

/**
 * A [SegmentTransport] that uploads to Google Drive (the production cloud relay) — the same receipt-wait
 * never-lose contract as [FileRelayTransport], but over Drive REST via [DriveClient]. `postPhoto` returns
 * `true` ONLY after the Mac's matching-`(token,epoch,group,seq,fp)` receipt appears, never on a write alone.
 * Adopts the Mac-published epoch from `_epoch.json` in the shared Drive folder.
 *
 * Uses query-or-update (find by `appProperties.relayName` → update, else create) so it stores NO fileId —
 * side-stepping the coexisting-duplicate hazard at the source (the Mac's reap is the backstop). Blocking
 * (called inside `withContext(Dispatchers.IO)`). Byte-format via [RelayObjectFormat]; parses the small
 * receipt/epoch objects with org.json. Kotlin mirror of the iOS `DriveRelayTransport`. `client`'s token
 * provider is a Google OAuth access token for the SAME account as the Mac (device sign-in supplies it).
 */
class DriveRelayTransport(
    private val client: DriveClient,
    private val token: String,
    private val receiptWaitTimeoutMs: Long = 20_000,
    private val receiptPollMs: Long = 500
) : SegmentTransport {

    private fun esc(s: String) = s.replace("'", "\\'")

    private fun folderId(): String? = try {
        client.listFiles("mimeType = 'application/vnd.google-apps.folder' and appProperties has { key='relayToken' and value='${esc(token)}' } and trashed = false").firstOrNull()?.id
    } catch (e: Exception) { null }

    private fun fileId(folder: String, name: String): String? = try {
        client.listFiles("'${esc(folder)}' in parents and appProperties has { key='relayName' and value='${esc(name)}' } and trashed = false").firstOrNull()?.id
    } catch (e: Exception) { null }

    private fun upsert(folder: String, name: String, data: ByteArray, mime: String) {
        try {
            val id = fileId(folder, name)
            if (id != null) client.updateMedia(id, data, mime)
            else client.createFile(name, listOf(folder), mapOf("relayName" to name, "relayToken" to token), data, mime)
        } catch (e: Exception) { /* best-effort: no receipt → phone's receipt-wait times out + retries; never a loss */ }
    }

    private fun read(folder: String, name: String): ByteArray? = try {
        val id = fileId(folder, name); if (id != null) client.getMedia(id) else null
    } catch (e: Exception) { null }

    private fun epoch(folder: String): String? {
        val d = read(folder, RelayObjectFormat.EPOCH_MARKER_NAME) ?: return null
        return try {
            val o = JSONObject(String(d, Charsets.UTF_8))
            if (o.optString("token") == token) o.optString("epoch").ifEmpty { null } else null
        } catch (e: Exception) { null }
    }

    private fun validReceipt(folder: String, group: String, seq: Int, epoch: String, fp: String): Boolean {
        val d = read(folder, RelayObjectFormat.receiptName(group, seq)) ?: return false
        return try {
            val o = JSONObject(String(d, Charsets.UTF_8))
            o.optString("kind") == "receipt" && o.optString("token") == token && o.optString("epoch") == epoch &&
                o.optString("group") == group && o.optString("seq") == seq.toString() && o.optString("fp") == fp
        } catch (e: Exception) { false }
    }

    override fun postPhoto(jpeg: ByteArray, group: String, seq: Int, type: String, priority: String?,
                           year: Int?, month: Int?, device: String, replaces: String?): Boolean {
        val yearS = year?.toString(); val monthS = month?.toString()
        val repl = if (!replaces.isNullOrEmpty()) replaces else null
        val fp = RelayObjectFormat.fingerprint(type, priority, yearS, monthS, repl)
        val deadline = System.currentTimeMillis() + receiptWaitTimeoutMs
        var folder: String? = null; var wroteForEpoch: String? = null
        do {
            if (folder == null) folder = folderId()
            if (folder == null) { Thread.sleep(1000); continue }          // Mac relay not up yet
            val f = folder!!
            val e = epoch(f)                                              // re-read each iteration (epoch may change on Mac restart)
            if (e == null) { Thread.sleep(1000); continue }               // no epoch yet
            if (validReceipt(f, group, seq, e, fp)) return true           // receipt-first
            if (wroteForEpoch != e) {                                     // write-once per epoch (re-write if epoch changes)
                upsert(f, RelayObjectFormat.jpegName(group, seq), jpeg, "image/jpeg")
                upsert(f, RelayObjectFormat.sidecarName(group, seq),
                    RelayObjectFormat.encodeSidecar(token, e, group, seq, type, priority, yearS, monthS, repl, device), "application/json")
                wroteForEpoch = e
            }
            Thread.sleep(receiptPollMs)
        } while (System.currentTimeMillis() < deadline)
        return false   // timeout → item stays FAILED → auto-retry re-enters (re-resolves epoch); local copy retained
    }

    override fun segmentComplete(group: String, priority: String?, year: Int?, month: Int?, seqs: String?): Boolean {
        val f = folderId() ?: return false; val e = epoch(f) ?: return false
        upsert(f, RelayObjectFormat.segmentName(group),
            RelayObjectFormat.encodeSegment(token, e, group, priority, year?.toString(), month?.toString(), seqs), "application/json")
        return true
    }

    override fun sessionComplete(): Boolean {
        val f = folderId() ?: return false; val e = epoch(f) ?: return false
        upsert(f, RelayObjectFormat.SESSION_COMPLETE_NAME, RelayObjectFormat.encodeSessionComplete(token, e), "application/json")
        return true
    }

    override fun sessionDisconnect(): Boolean = true
}

File: ArchiveProcessor/ArchiveCaptureiOS/Sources/ArchiveCaptureiOS/Net/DriveRelayTransport.swift
import Foundation

/// A `SegmentTransport` that uploads to Google Drive (the production cloud relay) — the same
/// receipt-wait contract as `FileRelayTransport`, but over Drive REST via `DriveClient`. `postPhoto`
/// returns `true` ONLY after the Mac's matching-`(token,epoch,group,seq,fp)` receipt appears, never on a
/// write alone (never-lose). Adopts the Mac-published epoch from `_epoch.json` in the shared Drive folder.
///
/// Uses **query-or-update** (find the file by `appProperties.relayName`, then update it, else create) — so
/// it never stores a fileId to lose, side-stepping the coexisting-duplicate hazard at the source (the Mac's
/// reap is the backstop). `client`'s token provider is a Google OAuth access token for the SAME account as
/// the Mac; the on-device sign-in that supplies it is the device-session follow-up. Tests inject a mock
/// `DriveClient`.
struct DriveRelayTransport: SegmentTransport {
    let client: DriveClient
    let token: String
    var receiptWaitTimeout: TimeInterval = 20
    var receiptPollInterval: TimeInterval = 0.5

    private func esc(_ s: String) -> String { s.replacingOccurrences(of: "'", with: "\\'") }

    /// The Mac creates the session folder + publishes `_epoch.json` on relay start; the phone only finds it.
    private func folderId() -> String? {
        (try? client.listFiles(query: "mimeType = 'application/vnd.google-apps.folder' and appProperties has { key='relayToken' and value='\(esc(token))' } and trashed = false"))?.first?.id
    }
    private func fileId(_ folder: String, _ name: String) -> String? {
        (try? client.listFiles(query: "'\(esc(folder))' in parents and appProperties has { key='relayName' and value='\(esc(name))' } and trashed = false"))?.first?.id
    }
    private func upsert(_ folder: String, _ name: String, _ data: Data, _ mime: String) throws {
        if let id = fileId(folder, name) { try client.updateMedia(fileId: id, media: data, mimeType: mime) }
        else { _ = try client.createFile(name: name, parents: [folder], appProperties: ["relayName": name, "relayToken": token], media: data, mimeType: mime) }
    }
    private func read(_ folder: String, _ name: String) -> Data? {
        guard let id = fileId(folder, name) else { return nil }
        return try? client.getMedia(fileId: id)
    }
    private func epoch(_ folder: String) -> String? {
        guard let d = read(folder, RelayObjectFormat.epochMarkerName), let m = RelayObjectFormat.parse(d), m["token"] == token else { return nil }
        return m["epoch"]
    }
    private func validReceipt(_ folder: String, group: String, seq: Int, epoch: String, fp: String) -> Bool {
        guard let d = read(folder, RelayObjectFormat.receiptName(group: group, seq: seq)), let m = RelayObjectFormat.parse(d) else { return false }
        return m["kind"] == "receipt" && m["token"] == token && m["epoch"] == epoch
            && m["group"] == group && m["seq"] == String(seq) && m["fp"] == fp
    }

    func postPhoto(jpeg: Data, group: String, seq: Int, type: String, quality: String?,
                   year: Int?, month: Int?, device: String, replaces: String?) async -> Bool {
        let yearS = year.map(String.init), monthS = month.map(String.init)
        let repl = (replaces?.isEmpty == false) ? replaces : nil
        let fp = RelayObjectFormat.fingerprint(type: type, quality: quality, year: yearS, month: monthS, replaces: repl)
        let deadline = Date().addingTimeInterval(receiptWaitTimeout)
        var folder: String?, wroteForEpoch: String?
        repeat {
            if folder == nil { folder = folderId() }                       // resolve once (Mac's folder)
            guard let f = folder else { try? await Task.sleep(nanoseconds: 1_000_000_000); continue }  // Mac relay not up
            guard let e = epoch(f) else { try? await Task.sleep(nanoseconds: 1_000_000_000); continue } // re-read each iteration (epoch may change on Mac restart)
            if validReceipt(f, group: group, seq: seq, epoch: e, fp: fp) { return true }                // receipt-first
            if wroteForEpoch != e {                                         // write-once per epoch (re-write if epoch changes)
                try? upsert(f, RelayObjectFormat.jpegName(group: group, seq: seq), jpeg, "image/jpeg")
                try? upsert(f, RelayObjectFormat.sidecarName(group: group, seq: seq),
                       RelayObjectFormat.encodeSidecar(token: token, epoch: e, group: group, seq: seq, type: type,
                           quality: quality, year: yearS, month: monthS, replaces: repl, device: device), "application/json")
                wroteForEpoch = e
            }
            try? await Task.sleep(nanoseconds: UInt64(receiptPollInterval * 1_000_000_000))
        } while Date() < deadline
        return false   // timeout → .failed → auto-retry re-enters (re-resolves epoch); local copy retained
    }

    func segmentComplete(group: String, quality: String?, year: Int?, month: Int?, seqs: String?) async -> Bool {
        guard let f = folderId(), let e = epoch(f) else { return false }
        do {
            try upsert(f, RelayObjectFormat.segmentName(group: group),
                   RelayObjectFormat.encodeSegment(token: token, epoch: e, group: group,
                       quality: quality, year: year.map(String.init), month: month.map(String.init), seqs: seqs), "application/json")
            return true
        } catch { return false }
    }
    func sessionComplete() async -> Bool {
        guard let f = folderId(), let e = epoch(f) else { return false }
        do {
            try upsert(f, RelayObjectFormat.sessionCompleteName, RelayObjectFormat.encodeSessionComplete(token: token, epoch: e), "application/json")
            return true
        } catch { return false }
    }
    func sessionDisconnect() async -> Bool { true }
}
